Sources/CoreMetrics/Metrics.swift

```diff
@@ -488,6 +580,74 @@ public protocol MetricsFactory {
+private class AccumulatingRoundingFloatingPointCounter: FloatingPointCounterHandler {
+    private static let kAutoStateResetThreshold = Double(Int64.max - (2 << 8))
+
+    private var lock = Lock()
+    private var rawValue: Double = 0
+    private var roundedValue: Int64 = 0
+    private(set) var counterHandler: CounterHandler
+
+    init(counterHandler: CounterHandler) {
+        self.counterHandler = counterHandler
+    }
+
+    func increment(by amount: Double) {
+        assert(amount < AccumulatingRoundingFloatingPointCounter.kAutoStateResetThreshold)
+        self.lock.withLockVoid {
+            let incremented = rawValue + amount
+            if incremented < AccumulatingRoundingFloatingPointCounter.kAutoStateResetThreshold {
+                self.rawValue = incremented
+            } else {
+                self.rawValue = amount
+                self.roundedValue = 0
+            }
+            let roundedIncrement = Int64(rawValue) - roundedValue
+            assert(roundedIncrement >= 0)
+            if roundedIncrement > 0 {
+                self.roundedValue += roundedIncrement
+                self.counterHandler.increment(by: roundedIncrement)
+            }
+        }
+    }
+
+    func reset() {
+        self.lock.withLockVoid {
+            self.rawValue = 0
+            self.roundedValue = 0
+            self.counterHandler.reset()
+        }
+    }
+}
+
+extension MetricsFactory {
+    /// Create a default backing `FloatingPointCounterHandler` for backends which do not naively support floating point counters.
+    ///
+    /// The created FloatingPointCounterHandler is a wrapper around a backend's CounterHandler which accumulates floating point increments and records increments to an underlying CounterHandler after crossing integer boundaries.
+    ///
+    /// - parameters:
+    ///     - label: The label for the `FloatingPointCounterHandler`.
+    ///     - dimensions: The dimensions for the `FloatingPointCounterHandler`.
+    public func makeFloatingPointCounter(label: String, dimensions: [(String, String)]) -> FloatingPointCounterHandler {
+        return AccumulatingRoundingFloatingPointCounter(
+            counterHandler: makeCounter(label: label, dimensions: dimensions))
+    }
+
+    /// Invoked when the corresponding `FloatingPointCounter`'s `destroy()` function is invoked.
+    /// Upon receiving this signal the factory may eagerly release any resources related to this counter.
+    ///
+    /// `destroyFloatingPointCounter` must be implemented if `makeFloatingPointCounter` is implemented.
+    ///
+    /// - parameters:
+    ///     - handler: The handler to be destroyed.
+    public func destroyFloatingPointCounter(_ handler: FloatingPointCounterHandler) {
+        if let handler = handler as? AccumulatingRoundingFloatingPointCounter {
+            destroyCounter(handler.counterHandler)
```

Comment:
handler.destroy and then handler.counterHandler can stay private?

Reply:
Are you thinking I should add a destroy method? The destroy instance funcs are only available on the `<Metric>` API not the `<Metrics>Handler` SPI, iiuc.

Reply:
right, since AccumulatingRoundingFloatingPointCounter is an internal type and we check the type here, we can presumably add a non-standard func to it?

Reply:
Well it will need access to the metrics factory so the method would end up looking like:
```swift
    func destroy(factory: MetricsFactory) {
        factory.destroyCounter(counterHandler)
    }

    public func destroyFloatingPointCounter(_ handler: FloatingPointCounterHandler) {
        if let handler = handler as? AccumulatingRoundingFloatingPointCounter {
            handler.destroy(factory: self)
        }
    }
```
I wasn't sure if we should pass the metrics factory to `AccumulatingRoundingFloatingPointCounter`; I'm happy to make this change if you think it makes sense.

Reply:
factory is sort of a s singleton, so maybe `MetricsSystem.factory.destroyCounter(self.counterHandler)` would work? none of them are too elegant, but at least with this the private state of FloatingPointCounterHandler is kept private

Reply:
Pushed this change